[Intel MKL] Add scripts for creating allreduce based Tensorflow k8s deployment

congxu-ml opened the pull request:
Create allreduce based distributed Tensorflow k8s deployment for testing

Intel optimized Tensorflow horovod dockerfile
- tensorflow/tools/docker/Dockerfile.devel-mkl-horovod
- tensorflow/tools/docker/Dockerfile.mkl-horovod

Signed-off-by: Cong Xu <cong.xu@intel.com>

congxu-ml committed:
[Intel MKL] Add scripts for creating allreduce based Tensorflow k8s deployment

Create allreduce based distributed Tensorflow k8s deployment for testing

Intel optimized Tensorflow horovod dockerfile
- tensorflow/tools/docker/Dockerfile.devel-mkl-horovod
- tensorflow/tools/docker/Dockerfile.mkl-horovod

Signed-off-by: Cong Xu <cong.xu@intel.com>

penpornk commented on tensorflow/tools/dist_test/README.md:
```diff
@@ -1,6 +1,6 @@
 # Testing Distributed Runtime in TensorFlow
-This folder containers tools and test suites for the GRPC-based distributed
-runtime in TensorFlow.
+This folder contains tools and test suites for the distributed runtime
```
`runtime` -> `runtimes`. Please explain that there are two types of distributed runtimes: GRPC-based and Allreduce-based. Please also add a reference to the Allreduce-based model (the Horovod paper is fine).

congxu-ml replied:
Sure. I added GRPC-based and Allreduce-based to that line, and added a reference to Horovod paper in "Create allreduce-based Tensorflow k8s deployment" section.

penpornk commented on tensorflow/tools/dist_test/README.md:
```diff
@@ -122,3 +122,32 @@ servers. For example:
 See [Kubernetes kubectl documentation](http://kubernetes.io/docs/user-guide/kubectl-overview/)
 for more details.
+
+**Create allreduce based Tensorflow k8s deployment**
```
Sorry for nitpicking. Could you please change all occurrences of `allreduce based` to `allreduce-based`?

congxu-ml replied:
I changed all allreduce based to allreduce-based.

penpornk reviewed (changes requested):
Thank you for your PR! I have minor formatting comments. I'm a little confused how this is related to Intel MKL, though.

congxu-ml commented on tensorflow/tools/dist_test/scripts_allreduce/k8s_deploy_tensorflow.sh:
```diff
@@ -0,0 +1,260 @@
+  echo "    num_containers: number of containers that will be launched."
+  echo "    deployment:     deployment name. (default: k8s-ml-deployment)"
+  echo "    configs_map:    configs map name. (default: k8s-configs-map)"
+  echo "    cp:             upload file to all containers."
```
I added "(src and dest must be provided along with cp option)" in the usage description.

congxu-ml commented on tensorflow/tools/dist_test/scripts_allreduce/k8s_deploy_tensorflow.sh:
```diff
@@ -0,0 +1,260 @@
+  echo "    deployment:     deployment name. (default: k8s-ml-deployment)"
+  echo "    configs_map:    configs map name. (default: k8s-configs-map)"
+  echo "    cp:             upload file to all containers."
+  echo "    src:            path to local source file."
```
I added "(used for cp option)" in the usage description.

congxu-ml commented on tensorflow/tools/dist_test/scripts_allreduce/k8s_deploy_tensorflow.sh:
```diff
@@ -0,0 +1,260 @@
+  echo "    port:           ssh port in container. Set ssh port (other than 22)"
+  echo "                    when host network mode is enabled"
+  echo "    hostnet:        enable host network mode. (default: disable)"
+  echo "    shared_volume:  Used to mount shared volume. (default: disable)"
```
Done

congxu-ml commented on tensorflow/tools/dist_test/scripts_allreduce/k8s_deploy_tensorflow.sh:
```diff
@@ -0,0 +1,260 @@
+
+# Python script to generate yaml file for k8s TensorFlow cluster
+CUR_SCRIPT_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
+K8S_GEN_AllREDUCE_TF_YAML="${CUR_SCRIPT_DIR}/k8s_generate_yaml.py"
```
Done

congxu-ml commented on tensorflow/tools/dist_test/scripts_allreduce/k8s_deploy_tensorflow.sh:
```diff
@@ -0,0 +1,260 @@
+# Used to mount shared volume
+USE_SHARED_VOLUME=0
+
+if [ $# -lt 1 ]; then
```
I changed [ to [[ for all ifs.

congxu-ml commented on tensorflow/tools/dist_test/scripts_allreduce/k8s_deploy_tensorflow.sh:
```diff
@@ -0,0 +1,260 @@
+  exit 1
+fi
+
+while [[ $# -ge 1 ]]
```
I put "do" on the same line with "while".

congxu-ml commented on tensorflow/tools/dist_test/scripts_allreduce/k8s_deploy_tensorflow.sh:
```diff
@@ -0,0 +1,260 @@
+
+function generate_container_hostfile
+{
+  sleep 3
```
I deleted "sleep 3" line.

congxu-ml commented on tensorflow/tools/dist_test/scripts_allreduce/k8s_deploy_tensorflow.sh:
```diff
@@ -0,0 +1,260 @@
+function generate_container_hostfile
+{
+  sleep 3
+  ${KUBECTL_BIN} get pods --output=wide | grep ${DEPLOYMENT} \
```
I added that comment.

penpornk replied:
Thank you!

congxu-ml commented on tensorflow/tools/dist_test/scripts_allreduce/k8s_deploy_tensorflow.sh:
```diff
@@ -0,0 +1,260 @@
+
+# Check kubectl binary
+KUBECTL_BIN=kubectl
+if ! [ -x "$(command -v ${KUBECTL_BIN})" ]; then
```
I moved ! inside square brackets.

congxu-ml commented on tensorflow/tools/dist_test/scripts_allreduce/k8s_deploy_tensorflow.sh:
```diff
@@ -0,0 +1,260 @@
+if [ $DELETE -eq 1 ]; then
+  echo "Deleting deployment ${DEPLOYMENT} and configs map ${CONFIGS_MAP}..."
+  delete_deployment_configmap
+elif [ $CP -eq 1 ] || [ -n "$SRC" ] || [ -n "$DEST" ] ; then
```
I merged them into a single bracket.

congxu-ml commented on tensorflow/tools/dist_test/scripts_allreduce/k8s_generate_yaml.py:
```diff
@@ -0,0 +1,94 @@
+DEFAULT_CONFIGS_MAP = 'k8s-configs-map'
+DEFAULT_DEPLOYMENT = 'k8s-ml-deployment'
+
+def main():
```
I added one line above main()

penpornk replied:
Thank you!

congxu-ml commented on tensorflow/tools/dist_test/scripts_allreduce/k8s_generate_yaml.py:
```diff
@@ -0,0 +1,94 @@
+def main():
+  """Do arg parsing."""
+  parser = argparse.ArgumentParser()
+  parser.register(
```
I removed it.

congxu-ml commented on tensorflow/tools/dist_test/scripts_allreduce/k8s_generate_yaml_lib.py:
```diff
@@ -0,0 +1,234 @@
+# machine that launches a TensorFlow k8s cluster does not have to have the
+# Python package of TensorFlow installed on it.
+
+COMFIGS_MAP = (
```
Done.

congxu-ml commented on tensorflow/tools/dist_test/scripts_allreduce/k8s_generate_yaml_lib.py:
```diff
@@ -0,0 +1,234 @@
+
+  CHANGE_SSH_PORT = get_change_ssh_port(use_hostnet, port)
+
+  config += COMFIGS_MAP.format(
```
Done.

congxu-ml commented on tensorflow/tools/dist_test/scripts_allreduce/k8s_generate_yaml_lib.py:
```diff
@@ -0,0 +1,234 @@
+
+  return CHANGE_SSH_PORT
+
+def get_hostnet(use_hostnet):
```
Done.

penpornk replied:
Thank you!

congxu-ml commented on tensorflow/tools/dist_test/scripts_allreduce/k8s_deploy_tensorflow.sh:
```diff
@@ -0,0 +1,260 @@
+# ==============================================================================
+
+function usage
+{
```
Done

congxu-ml commented on tensorflow/tools/dist_test/README.md:
```diff
@@ -122,3 +122,32 @@ servers. For example:
+        --num_containers <num_of_containers> \
+        --image <docker_image> \
+        --deployment <deployment_name> \
+        --configs_map <configs_map>
```
Sure

congxu-ml commented:
Thank you so much for your great review! I have addressed all the comments, please have a look at my new  commit. We use [Intel MKL] to track all the PRs submitted by Intel, it doesn't necessarily have to be related to Intel MKL code.

penpornk commented on tensorflow/tools/dist_test/scripts_allreduce/k8s_deploy_tensorflow.sh:
```diff
@@ -0,0 +1,254 @@
+  echo "Deleting deployment ${DEPLOYMENT} and config map ${CONFIG_MAP}..."
+  delete_deployment_configmap
+elif [[ $CP -eq 1 || -n "$SRC" || -n "$DEST" ]] ; then
+  if [[ "$CP" -eq 1 && -n "$SRC"  && -n "$DEST" ]]; then
```
```suggestion
  if [[ "$CP" -eq 1 && -n "$SRC" && -n "$DEST" ]]; then
```

congxu-ml replied:
Done

penpornk commented on tensorflow/tools/dist_test/scripts_allreduce/k8s_generate_yaml_lib.py:
```diff
@@ -0,0 +1,225 @@
+
+def get_hostnet(use_hostnet):
+  if use_hostnet == 1:
+    return """\n\
```
Triple quotes preserve newlines in strings, so I think we can remove `\n\`s?

congxu-ml replied:
Yes, you are right! We can remove \n\s because triple quotes can span multiple lines and preserve newlines in the strings.

penpornk reviewed (changes requested):
Thank you very much for the changes! I have a few very, very minor edits and we should be good to go.
[Intel MKL] tag: Got it. :)

congxu-ml commented on tensorflow/tools/dist_test/README.md:
```diff
@@ -122,3 +122,37 @@ servers. For example:
+https://arxiv.org/pdf/1802.05799.pdf.
+
+The script "scripts_allreduce/k8s_deploy_tensorflow.sh" can be used to create
+or delete a allreduce-based Tensorflow k8s deployment with specified number of
```
Done

congxu-ml commented:
Thanks a lot for the great suggestions, they are very useful! I have addressed your comments, please have a look at my new commit. :)

penpornk reviewed (dismissed):
Thanks again for the PR!

congxu-ml commented:
The commit does not pass Ubuntu Santiy check because line 64 in k8s_generate_yaml.py is too long (84/88). I am going to modify that line and see whether the new commit can pass Ubuntu Santiy check.

congxu-ml commented:
The commit does not pass Ubuntu Santiy check because line 64 in k8s_generate_yaml.py is too long (84/88). I have modified that line and submitted a new commit. Please approve and new commit and see whether it can pass Ubuntu Santiy check. Thanks!

penpornk reviewed (approved):
Thank you for looking into this!

congxu-ml commented:
The new commit passes the Ubuntu Santiy check. Thanks a lot!
